src/crypto.js

```
@@ -96,47 +112,82 @@ const hash256 = async (msg, type = 'SHA-256') => {
+ * Decrypt the encypted MK
+ * return the raw MK
+ *
+ * @param {string | arrayBuffer} passPhrase The passphrase that is used to derive the key
+ * @param {arrayBuffer} [salt] The salt
+ * @param {Number} [iterations] The iterations number
+ * @param {string} [hash] The hash function used for derivation and final hash computing
+ * @param {encMK} [encMK] The encrypted MK
+ * @returns {Promise<Uint8Array>}   A promise that contains the hashed derived key
+ */
+const deriveBitsDecMK = async (passPhrase, salt, iterations, hash, encMK) => {
+  const _salt = typeof (salt) === 'string' ? Buffer.from(salt, ('hex')) : salt
+  const derivedPassphrase = await deriveBits(passPhrase, _salt, iterations, hash)
+  const KEK = await importKey(derivedPassphrase)
+  return decrypt(KEK, encMK)
+}
+
+const requiredParameterProtectedMK = ['salt', 'iterations', 'encMK', 'hashAlgo']
 
 /**
  * Check a given passphrase by comparing it to the stored hash value (in HashedPassphrase object)
  *
  * @param {string} passphrase The passphrase
- * @param {HashedPassphrase} hashedPassphrase The HashedPassphrase object
+ * @param {protectedMK} protectedMK The protectedMK object
  * @returns {Promise<Boolean>}   A promise
  */
-const checkPassphrase = async (passPhrase, hashedPassphrase) => {
+const checkPassphrase = async (passPhrase, protectedMK) => {
   _checkPassphrase(passPhrase)
-  checkObject(hashedPassphrase, requiredParameterHashedPassphrase)
-  const { salt, iterations, storedHash, hashAlgo } = hashedPassphrase
-  const hashCandidate = await deriveBitsAndHash(passPhrase, Buffer.from(salt, 'hex'), iterations, hashAlgo)
-  return Buffer.from(hashCandidate).toString('hex') === storedHash
+  checkObject(protectedMK, requiredParameterProtectedMK)
+  try {
+    const { salt, iterations, encMK, hashAlgo } = protectedMK
+    const MK = await deriveBitsDecMK(passPhrase, salt, iterations, hashAlgo, encMK)
+    return Buffer.from(MK, 'hex')
+  } catch (error) {
```

why not let the error be thrown like before ? Masq-app will display an error if this method throws, it was the right thing to do I think, wdyt ?

---

because we want to return null if the passphrase is wrong and the MK directly (if right passphrase)